<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %><jsp:include page="/includes/header.jsp" /><table width="100%" class="table_back" >  <tr>    <td align="left"><img name="help_nav" src="fireworks/images/help_nav.gif" width="600" height="18" border="0" usemap="#m_help_nav" alt="">      <map name="m_help_nav">        <area shape="rect" coords="49,1,89,18" href="faq.htm" alt="" >        <area shape="rect" coords="106,1,155,18" href="pages.htm" alt="" >        <area shape="rect" coords="173,1,247,18" href="../index.html" alt="" >        <area shape="rect" coords="-1,-2,34,18" href="index.htm" alt="" >      </map>    </td>  </tr></table><table width="95%" align="center" cellspacing="5" class="table_back">  <tr class="help_h2">    <td>      <h2>Frequently Asked Questions </h2>    </td>  </tr>  <tr class="help_h2">    <td><p><strong><a name="One"></a>What type of data does the <c:out value="${proj}" /> database hold? </strong></p>      <p>The <c:out value="${proj}" /> database currently holds In-Situ Hybridization <c:if test="${proj == 'GUDMAP'}">and Microarray</c:if> gene expression results from experiments carried out on the <c:out value="${perspective}" /> system of  mice aged TS17 and above.</p>      <p><strong><a name="Two"></a>Why can I only see one thumbnail image when I submitted more than one?</strong></p>      <p>In the <a href="ish_browse_all.htm">Browse ISH Data</a> result set, the thumbnail is in the last column. The width of the table is as set size and so there is only room to display the one image. The image displayed is the first one referenced in the submitted data. To see the other images associated with the submission, click on the <c:out value="${proj}" />:ID to display the <a href="ish_submission.htm">Submission Details</a> page and all the images are found under the 'Original Images' section. </p>      <p><strong><a name="Three"></a>What is a Collection and how can I create one?</strong></p>      <p>A collection is a personal set of entries in the database in which the user is interested. For details of how to create a collection see <a href="collections.htm#Ish">ISH collections</a><c:if test="${proj == 'GUDMAP'}"> or <a href="collections.htm#Microarray">Microarray collections </a></c:if></p>      <p><strong><a name="Four"></a>Can I share my Collection with other users?</strong></p>      <p>Not at present but we are working on a facility which will allow users to:</p>      <ul>        <li>Email the collection to other users</li>        <li>Save the collection locally</li>        <li>Upload a saved collection</li>      </ul>      <p><strong><a name="Five"></a>What is the difference between 'Available' and 'Editing in Progress'? </strong></p>      <ul>        <li>'Available' refers to the number of entries in the database which have been approved by the editorial office and are available for public viewing. </li>        <li>'Editing in Progress' refers to the total number of submissions sent to the Database group by a particular lab on a particular date MINUS the number 'Available' for that same data set. </li>      </ul>      <c:if test="${proj == 'GUDMAP'}">      <p><strong><a name="Six"></a>How do I query micro-array data?&nbsp; </strong></p>      <p>Select '<a href="mic_browse_all.htm">Browse Microarray Data</a>' submenu from the 'Browse All ' menu. All the results returned from the query page only return ISH data at present. This will be extended to inclued Microrarray data. </p>      <p><strong><a name="Seven"></a>How do I identify the gene I am interested in from the gene list returned from a micro-array submission?</strong></p>      <p>Select '<a href="mic_browse_all.htm">Browse Microarray Data</a>' submenu from the 'Browse All ' menu. Click on a <c:out value="${proj}" />:ID on the results page. This will take you to the Microarray Submission Details page. The gene list section provides a sortable gene symbol column. Click on a gene symbol to display further information. </p>      <p><strong><a name="Eight"></a>Can I get a list of genes (as gene symbols) that are called present from a micro-array entry in the <c:out value="${proj}" /> database?</strong></p>      <p>Yes. Select '<a href="mic_browse_all.htm">Browse Microarray Data</a>' submenu from the 'Browse All ' menu. Click on a <c:out value="${proj}" />:ID on the results page. The gene list section provides a sortable detection column.</p>      <p><strong><a name="Nine"></a>Can I download the array data to analyse myself?</strong></p>      <p>Yes, you can do this is 2 ways:</p>      <ol>        <li>Select '<a href="mic_browse_all.htm">Browse Microarray Data</a>' submenu from the 'Browse All ' menu. Click on a <c:out value="${proj}" />:ID on the results page. Raw data and other supplementary files are downloadable in the 'Supplementary Data Files' section.</li>        <li>If you are a <c:out value="${proj}" /> consortium member you can download all submitted data from the <a href="http://aberlour.hgu.mrc.ac.uk/gudmap_ftp"><c:out value="${proj}" /> ftp site</a>. </li>      </ol>      <p><strong><a name="Ten"></a>Can I download ISH data?&nbsp; </strong></p>      <p>If you are a GUDMAP consortium member you can download all submitted data from the <a href="http://aberlour.hgu.mrc.ac.uk/gudmap_ftp">GUDMAP ftp site</a>.</p>      <p>Discussion is underway to agrree on an export format for this data. </p>      </c:if>      <p><strong><a name="Eleven"></a>I have a gene expression database, can I&nbsp; get hold of <c:out value="${proj}" /> data in database-ready form to enter directly into my database?</strong></p>      <p>This is still under discussion. It may be that the data is made available as a database dump or output to spreadsheet.</p>      <p><strong><a name="Twelve"></a>How can I find results for a named gene (e.g. Wnt4) if it doesn&rsquo;t come up in the query &lsquo;Which components express a gene? &lsquo; on the query page? </strong></p>      <p>If there is an entry for Wnt4 then 'Browse ISH Data' should pick it up. Use a combination of gene symbol sort and 'go to page' to find the entry. It is proposed to implement a refinement of the browse all list to go to a particular page which contains a gene symbol.</p>      <p>The reason the entry may not have shown up from the query page is that the entry may not have been annotated. Annotation must be present to ask 'Which components express a gene?' </p>      <c:if test="${proj == 'GUDMAP'}">      <p><strong><a name="Thirteen"></a>Does the query page query both array and ISH data?</strong></p>      <p>No. You can only make ISH queries from the query page. Work is underway to include microarray data in searches made from the query page. </p>      </c:if>      <p><strong><a name="Fourteen"></a>Can I make a query using a list of genes?</strong></p>      <p>No. Work is underway to make this possible either by uploading a text file or typing in a list of genes to the interface. </p>      <p><strong><a name="Fifteen"></a>How can I contact the person who generated a particular result?&nbsp; </strong></p>      <p>Click on a <c:out value="${proj}" />:ID. In the resultant page there is an email address for the Submitter of the data.</p>      <p><strong><a name="Sixteen"></a>Can I get hold of the probe that was used?</strong></p>      <p>Click on a <c:out value="${proj}" />:ID. In the resultant page there is an email address for the Submitter of the data. You should contact the submitter and make a request. </p>      <p><strong><a name="Seventeen"></a>Where does the data on the gene page come from and is it updated periodically?</strong></p>      <p>The data for gene symbol, name, synonym and MGI link  comes from MGI. </p>      <p>The data for Chromosome details and Ensembl link comes from Ensembl.</p>      <p>Both the above are updated weekly.</p>      <p>The other external reference links are static URLs with the gene symbol embedded.</p>      <p>The remainder of the data is from the <c:out value="${proj}" /> database. </p>      <c:if test="${proj == 'GUDMAP'}">      <p><strong><a name="Eighteen"></a>Can I do any analyses on microarray data? </strong></p>      <p>Yes, there is a prototype analysis viewer available by clicking on the Future Devlopments link in the footer of each page and clicking on Microarray Visualisation. Go there directly via this link: <a href="http://gudmap.hgu.mrc.ac.uk/gudmap/arrayanalysis">http://gudmap.hgu.mrc.ac.uk/gudmap/arrayanalysis</a></p>      <p>A user guide to the analysis tool is available <a href="mic_analysis.htm">here</a>: </p>      </c:if>      <p>&nbsp;</p>    </td>  </tr></table><p class ="plaintextbold">&nbsp;</p><jsp:include page="/includes/footer.jsp" />